### PR TITLE
Add route53 plugin

### PR DESCRIPTION
**Pull request by yongtang**
### 1. What does this pull request do?

This fix adds route53 plugin so that it is possible to query route53 record through CoreDNS.

### 2. Which issues (if any) are related?

This fix is partially related to #1330

### 3. Which documentation changes (if any) need to be made?

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by miekg** [comment]
Should prolly keep this in a coredns/route53 repo? 
We have 33/34? plugins at the moment....

File: plugin/route53/route53.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+type Route53 struct {
+	Next plugin.Handler
+
+	zones  []string
+	keys   map[string]string
+	client route53iface.Route53API
+}
+
+// ServeDNS implements the plugin.Handler interface.
+func (route Route53) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
```
Comment (miekg) [inline plugin/route53/route53.go]: receivers are usually 1 or 2 chars, `r` is enouh

File: plugin/route53/setup.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+				return route53.New(session.Must(session.NewSession(&aws.Config{
+					Credentials: credential,
+				})))
+			}
+			return setupRoute53(c, f)
+		},
+	})
+}
+
+func setupRoute53(c *caddy.Controller, f func(*credentials.Credentials) route53iface.Route53API) error {
```
Comment (miekg) [inline plugin/route53/setup.go]: Just `setup`

**Comment by yongtang** [comment]
@miekg The PR has been updated with review comments addressed.

Eventually we may have to find a good way to manage all those plugins. Though for cloud integrations like` route53` and `google cloud dns` are quite important. Those help k8s deployments with "hybrid-cloud", which aligns well with CNCF/k8s strategy.

For a hybrid cloud, it is a lot easier to retrieve the DNS record through (AWS/Google Cloud) API directly than route the DNS traffic to a specific IP address (may change from time to time) inside AWS/Google Cloud. One scenario is the server deployed onprem or in google cloud to access DNS in AWS.

Maybe we could add this plugin for now? I think I could spend some time looking at some ideas on how to manage all the 30 plugins.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Add route53 p..." ]
>@miekg The PR has been updated with review comments addressed.
>
>Eventually we may have to find a good way to manage all those plugins. Though for cloud integrations like` route53` and `google cloud dns` are quite important. Those help k8s deployments with "hybrid-cloud", which aligns well with CNCF/k8s strategy.
>
>For a hybrid cloud, it is a lot easier to retrieve the DNS record through (AWS/Google Cloud) API directly than route the DNS traffic to a specific IP address (may change from time to time) inside AWS/Google Cloud. One scenario is the server deployed onprem or in google cloud to access DNS in AWS.
>
>Maybe we could add this plugin for now? I think I could spend some time looking at some ideas on how to manage all the 30 plugins.

I agree, this one is good to have.

**Pull request merged**